File: Services/HubSpotService.cs
using System.Net.Http.Headers;
using System.Text.Json;

namespace Api.Services
{
    public class HubSpotService
    {
        private readonly HttpClient _httpClient;
        private readonly string _apiKey;
        private static readonly JsonSerializerOptions _jsonOptions = new()
        {
            PropertyNameCaseInsensitive = true
        };

        public HubSpotService(HttpClient httpClient, IConfiguration configuration)
        {
            _httpClient = httpClient;
            _apiKey = configuration["HubSpot:ApiKey"] ?? throw new ArgumentNullException(nameof(configuration), "HubSpot:ApiKey is missing in configuration");

            _httpClient.BaseAddress = new Uri("https://api.hubapi.com/");
            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
            _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {_apiKey}");
        }


        public async Task<IEnumerable<HubSpotCustomer>> GetCustomers(string? after = null)
        {
            try
            {
                var url = "crm/v3/objects/contacts?limit=10";

                if (!string.IsNullOrEmpty(after))
                {
                    url += $"&after={after}";
                }

                var response = await _httpClient.GetAsync(url);
                response.EnsureSuccessStatusCode();

                var content = await response.Content.ReadAsStringAsync();
                var customers = JsonSerializer.Deserialize<HubSpotResponse>(content, _jsonOptions);

                return customers?.Results ?? [];
            }
            catch (JsonException jsonEx)
            {
                throw new Exception("Failed to deserialize the response from HubSpot", jsonEx);
            }
            catch (Exception ex)
            {
                throw new Exception("Failed to get customers from HubSpot", ex);
            }
        }
    }

    public record HubSpotResponse
    {
        public required IEnumerable<HubSpotCustomer> Results { get; set; }
    }

    public record HubSpotCustomer
    {
        public required string Id { get; set; }
        public required HubSpotCustomerProperties Properties { get; set; }
    }

    public record HubSpotCustomerProperties
    {
        public DateTime UpdatedAt { get; set; }
        public required string FirstName { get; set; }
        public required string LastName { get; set; }
        public required string Email { get; set; }
    }
}

File: services/HubSpotService.go
package services

import (
	"api/config"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// HubSpotService handles integration with HubSpot API.
type HubSpotService struct {
	client  *http.Client
	apiKey  string
	baseURL string
}

func GetHubSpotService() *HubSpotService {
	apiKey := config.Envs.HubSpotApiKey

	httpClient := &http.Client{
		Timeout: 10 * time.Second,
	}

	return &HubSpotService{
		client:  httpClient,
		apiKey:  apiKey,
		baseURL: "https://api.hubapi.com/",
	}
}

type HubSpotCustomer struct {
	ID         string                 `json:"id"`
	Properties HubSpotCustomerDetails `json:"properties"`
}

type HubSpotCustomerDetails struct {
	UpdatedAt time.Time `json:"updatedAt"`
	FirstName string    `json:"firstName"`
	LastName  string    `json:"lastName"`
	Email     string    `json:"email"`
}

type HubSpotCustomerResponse struct {
	Results []HubSpotCustomer `json:"results"`
}

func (s *HubSpotService) GetCustomers(after string) ([]HubSpotCustomer, error) {
	url := fmt.Sprintf("%scrm/v3/objects/contacts?limit=10", s.baseURL)
	if after != "" {
		url += fmt.Sprintf("&after=%s", after)
	}

	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+s.apiKey)
	req.Header.Set("Accept", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch customers, status code: %d", resp.StatusCode)
	}

	var hubSpotResponse HubSpotCustomerResponse
	if err := json.NewDecoder(resp.Body).Decode(&hubSpotResponse); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return hubSpotResponse.Results, nil
}
